worker-message/src/worker.rs

```diff
@@ -0,0 +1,325 @@
+            let coordinator_msg: CoordinatorMessage =
+                bincode::deserialize(&buffer).expect("Failed to deserialize CoordinatorMessage");
+            tx.send(coordinator_msg).await.ok(/*Only fail due to process exiting*/);
+        }
+    });
+    let (coordinator_tx, mut rx) = mpsc::channel(8);
+    tasks.spawn(async move {
+        let mut output = tokio::io::stdout();
+        loop {
+            let worker_msg = rx.recv().await.expect("Failed to write to standard output");
+            info!("{:#?}", worker_msg);
+            let encoded =
+                bincode::serialize(&worker_msg).expect("Failed to serialize WorkerMessage");
+            output
+                .write_u64(encoded.len() as u64)
+                .await
+                .expect("Failed to write to standard output");
+            output
+                .write_all(&encoded)
+                .await
+                .expect("Failed to write to standard output");
+            output
+                .flush()
+                .await
+                .expect("Failed to flush standard output");
+        }
+    });
+
+    tasks.spawn(async move {
+        let time_in_nanos = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap() // now can't be earlier than UNIX_EPOCH.
+            .as_nanos();
+        let project_dir = TempDir::new(&format!("playground-{time_in_nanos}")).expect("Failed to create temporary project directory");
+        // TODO: may change this to a hashmap to allow multiple concurrent jobs.
+        let mut current_job: Option<JoinHandle<()>> = None;
+        let mut current_stdin_tx = None;
+
+        let project_path = project_dir.path();
+        loop {
+            let coordinator_msg = coordinator_rx.recv().await.expect("Failed to receive CoordinatorMessage from deserialization task");
+            info!("{:#?}", coordinator_msg);
+            match coordinator_msg {
+                CoordinatorMessage::Request(job_id, job) => {
+                    if let Some(current_job) = current_job {
+                        current_job.abort();
+                    }
+
+                    let coordinator_tx = coordinator_tx.clone();
+                    let project_path = project_path.to_path_buf();
+
+                    let task_handle = if job.is_stream() {
+                        let (stdin_tx, stdin_rx) = mpsc::channel(8);
+                        current_stdin_tx = Some(stdin_tx);
+                        tokio::spawn(async move {
+                            let mut reqs = job.reqs;
+                            // is_stream() ensures it's nonempty.
+                            let stream_req = reqs.pop().unwrap();
+                            let stream_req = if let Request::StreamCommand(req) = stream_req {
+                                req
+                            } else {
+                                unreachable!()
+                            };
```

Comment:
`is_stream()` is determined by whether the last command is `StreamCommand`.

Maybe we can make `StreamCommand` a struct rather than a operation enum variant and add it as an additional field `Option<StreamCommand>` to `Job`.

Seems more elegant than current solution.

Reply:
I do not understand why `StreamCommand` is treated separately. I'd expect that basically all commands to be of the same form, roughly


```rust
loop {
    select! {
        Some(msg) = from_coordinator.recv() => {
            match msg {
                WriteFile(args) => spawn(do_write_file(args, to_coordinator)),
                RunProgram(args) => spawn(do_run_program(args, to_coordinator)),
                // etc.
            }
        }
    }
}
```

`RunProgram` is likely more complicated, as we might want multiple concurrent programs and would want to direct stdin to a specific program. More handwaving:

```rust
match msg {
    RunProgram(key, args) => {
        programs[key].abort();
        programs[key] = spawn(do_run_program(args, to_coordinator));
    },
    WriteStdin(key, args) => {
        programs[key].stdin.send(args);
    }
}
```

Reply:
`StreamCommand` pipes stdin/out/err from/to coordinator and then to frontend.
It requires a `Receiver<StdinMessage>` which is unique under `mpsc::channel`.

But of course we can change to support multiple streaming programs by attaching key(id) to each stdin segment.

Reply:
I think my point is mostly that there shouldn't be a separate `StreamCommand` — all commands should be streaming. Perhaps another way of looking at this:

```rust
struct RunCommand {
    unique_id: u32,
    key: String,
    program_name: String,
    // And all the other details needed
}

struct CommandOutput {
    unique_id: u32,
    kind: StdoutOrStderr,
    value: String,
}

async fn worker_main_logic() {
    let (process_tx, process_rx) = mpsc::channel();
    tokio::spawn(manage_processes(coordinator_tx.clone(), process_rx));

    while let Some(command) = coordinator_rx.recv().await {
        match command {
            Message::RunCommand(cmd) => process_tx.send(cmd)
        }
    }
}

async fn manage_processes(coordinator_tx: (), process_rx: ()) {
    let mut commands = HashMap::new();

    // Probably actually a `loop` and `select!` so we can watch our
    // children spawned tasks.
    while let Some(cmd) = process_rx.recv().await {
        let RunCommand { unique_id, key, .. } = cmd;
        
        if let Some(old_command) = commands.remove(&key) {
            do_proper_shutdown_stuff();
        }
        
        let mut spawned = process::spawn(cmd.program_name);
        let stdout = spawned.stdout.take();
        let stderr = spawned.stderr.take();

        let (stdout_complete_tx, stdout_complete_rx) = oneshot::channel::<()>();
        let (stderr_complete_tx, stderr_complete_rx) = oneshot::channel::<()>();
        
        tokio::spawn(async {
            // wait for process to exit
            spawned.wait().await;
            // wait for output to be sent
            join!(stdout_complete_rx, stderr_complete_rx);
            // tell coordingator process is done
            coordinator_tx.send(CommandComplete { unique_id })
        });

        // Copy stdout to coordinator
        tokio::spawn(async move {
            while Ok(value) = read_line(stdout).await {
                coordinator_tx.send(CommandOutput { unique_id, kind: Stdout, value })
            }
            
            stderr_complete_rx.send(());
        });
        // Repeat for stderr

        commands.insert(key, things_needed_to_shutdown_early);
    }
}
```